src/udf/python/risingwave/udf.py

```diff
@@ -392,11 +393,16 @@ def do_exchange(self, context, descriptor, reader, writer):
             raise e
 
     def serve(self):
-        """Start the server."""
+        """
+        Block until the server shuts down.
+
+        This method only returns if shutdown() is called or a signal (SIGINT, SIGTERM) received.
+        """
         print(
             "Note: You can use arbitrary function names and struct field names in CREATE FUNCTION statements."
             f"\n\nlistening on {self._location}"
         )
+        signal.signal(signal.SIGTERM, lambda s, f: self.shutdown())
```

Comment:
New requests will be rejected after the server is shutting down, right? 🤔️ And we will retry in the UDF client. Do I understand correctly?

Reply:
Yes. But the retry has not yet implemented in CN. We will fix that soon. 😅